lib/elixir/lib/module/types/descr.ex

```diff
@@ -1066,6 +1078,308 @@ defmodule Module.Types.Descr do
+    else
+      check_elements([], tag, elements, neg_elements, negs) and
+        check_compatibility(n, m, tag, elements, neg_tag, negs)
+    end
+  end
+
+  # Recursively check elements for emptiness
+  defp check_elements(_, _, _, [], _), do: true
+
+  defp check_elements(acc, tag, elements, [neg_type | neg_elements], negs) do
+    {ty, elements} = List.pop_at(elements, 0, term())
+    diff = difference(ty, neg_type)
+
+    (empty?(diff) or tuple_empty?(tag, Enum.reverse(acc, [diff | elements]), negs)) and
+      check_elements([ty | acc], tag, elements, neg_elements, negs)
+  end
+
+  # Determines if the set difference is empty when:
+  # - Positive tuple: {tag, elements} of size n
+  # - Negative tuple: open or closed tuples of size m
+  defp check_compatibility(n, m, tag, elements, neg_tag, negs) do
+    # The tuples to consider are all those of size n to m - 1, and if the negative tuple is
+    # closed, we also need to consider tuples of size greater than m + 1.
+    tag == :closed or
+      (Enum.all?(n..(m - 1)//1, &tuple_empty?(:closed, tuple_fill(elements, &1), negs)) and
+         (neg_tag == :open or tuple_empty?(:open, tuple_fill(elements, m + 1), negs)))
+  end
+
+  @doc """
+  Fetches the type of the value returned by accessing `index` on `tuple`
+  with the assumption that the descr is exclusively a tuple (or dynamic).
+
+  Returns one of:
+  - `{false, type}` if the element is always accessible and has the given `type`.
+  - `{true, type}` if the element may exist and has the given `type`.
+  - `:badindex` if the index is never accessible in the tuple type.
+  - `:badtuple` if the descr is not a tuple type.
+
+  ## Examples
+
+      iex> tuple_fetch(tuple([integer(), atom()]), 0)
+      {false, integer()}
+
+      iex> tuple_fetch(tuple([integer(), atom()]), 2)
+            :badindex
+
+      iex> tuple_fetch(union(tuple([integer()]), tuple([integer(), atom()])), 1)
+      {true, atom()}
+
+      iex> tuple_fetch(dynamic(), 0)
+      {true, dynamic()}
+
+      iex> tuple_fetch(integer(), 0)
+          :badtuple
+
+  """
+  def tuple_fetch(_, index) when index < 0, do: :badindex
+  def tuple_fetch(:term, _key), do: :badtuple
+
+  def tuple_fetch(%{} = descr, key) do
+    case :maps.take(:dynamic, descr) do
+      :error ->
+        if is_map_key(descr, :tuple) and tuple_only?(descr) do
+          {static_optional?, static_type} =
+            tuple_fetch_static(descr, key) |> pop_optional_static()
+
+          cond do
+            empty?(static_type) -> :badindex
+            static_optional? -> {true, static_type}
+            true -> {false, static_type}
+          end
+        else
+          :badtuple
+        end
+
+      {:term, static} ->
+        if tuple_only?(static) do
+          static_type = tuple_fetch_static(static, key)
+          {true, union(dynamic(), static_type)}
+        else
+          :badtuple
+        end
+
+      {%{map: {:open, [], []}}, static} when static == @none ->
```

Comment:
I guess we don't want to check for map here? :)

```suggestion
      {%{tuple: {:open, [], []}}, static} when static == @none ->
```